actors/builtin/miner/miner_actor.go

```diff
@@ -1630,33 +1646,47 @@ func handleProvingDeadline(rt Runtime) {
 
 			powerDelta = powerDelta.Sub(newFaultyPower)
-			penalizePowerTotal = big.Sum(penalizePowerTotal, newFaultyPower.QA, failedRecoveryPower.QA)
-
-			// Unlock sector penalty for all undeclared faults.
-			penaltyTarget := PledgePenaltyForUndeclaredFault(epochReward.ThisEpochRewardSmoothed, pwrTotal.QualityAdjPowerSmoothed,
-				penalizePowerTotal, rt.NetworkVersion())
-			// Subtract the "ongoing" fault fee from the amount charged now, since it will be added on just below.
-			penaltyTarget = big.Sub(penaltyTarget, PledgePenaltyForDeclaredFault(epochReward.ThisEpochRewardSmoothed, pwrTotal.QualityAdjPowerSmoothed, penalizePowerTotal))
-			penaltyFromVesting, penaltyFromBalance, err := st.PenalizeFundsInPriorityOrder(store, currEpoch, penaltyTarget, unlockedBalance)
-			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to unlock penalty")
-			unlockedBalance = big.Sub(unlockedBalance, penaltyFromBalance)
-			penaltyTotal = big.Sum(penaltyTotal, penaltyFromVesting, penaltyFromBalance)
-			pledgeDelta = big.Sub(pledgeDelta, penaltyFromVesting)
 
+			if networkVersion >= network.Version3 {
+				// From network version 3, faults detected from a missed PoSt pay nothing.
+				// Failed recoveries pay nothing here, but will pay the ongoing fault fee in the subsequent block.
+			} else {
+				penalizePowerTotal := big.Add(newFaultyPower.QA, failedRecoveryPower.QA)
+
+				// Unlock sector penalty for all undeclared faults.
+				penaltyTarget := PledgePenaltyForUndeclaredFault(epochReward.ThisEpochRewardSmoothed, pwrTotal.QualityAdjPowerSmoothed,
+					penalizePowerTotal, rt.NetworkVersion())
+				// Subtract the "ongoing" fault fee from the amount charged now, since it will be added on just below.
+				penaltyTarget = big.Sub(penaltyTarget, PledgePenaltyForDeclaredFault(epochReward.ThisEpochRewardSmoothed,
+					pwrTotal.QualityAdjPowerSmoothed, penalizePowerTotal, networkVersion))
+				penaltyFromVesting, penaltyFromBalance, err := st.PenalizeFundsInPriorityOrder(store, currEpoch, penaltyTarget, unlockedBalance)
+				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to unlock penalty")
+				unlockedBalance = big.Sub(unlockedBalance, penaltyFromBalance)
+				penaltyTotal = big.Sum(penaltyTotal, penaltyFromVesting, penaltyFromBalance)
+				pledgeDelta = big.Sub(pledgeDelta, penaltyFromVesting)
+			}
 		}
 		{
 			// Record faulty power for penalisation of ongoing faults, before popping expirations.
 			// This includes any power that was just faulted from missing a PoSt.
-			penaltyTarget := PledgePenaltyForDeclaredFault(epochReward.ThisEpochRewardSmoothed, pwrTotal.QualityAdjPowerSmoothed, deadline.FaultyPower.QA)
+			ongoingFaultyPower := deadline.FaultyPower.QA
+			if networkVersion >= network.Version3 {
+				// From network version 3, this *excludes* any power that was just faulted from missing a PoSt.
+				ongoingFaultyPower = previouslyFaultyPower
```

Comment:
@lucaniz @irenegia I looked over this again and it appears to be correct.  `previouslyFaultPower` already accounts for `failedRecoveryPower` (we only handle failedRecoveryPower separately in the previous version because we used to charge this power SP).  Another way to put it, the value of `deadline.FaultyPower.QA` at line 1683 after processing deadline end = `previouslyFaultPower - newFaultyPower`